### PR TITLE
[pytorch] Add native support for segment reduce step1: API definition

### PR DESCRIPTION
**Pull request by serhaty**
Summary:
This is first diff to add native support for segment reduction in PyTorch. It provides similar functionality like torch.scatter or "numpy.ufunc.reduceat".

This diff mainly focuses on API layer to make sure future improvements will not cause backward compatibility issues. Once API is settled, here are next steps I am planning:
- Add support for other major reduction types (e.g. min, sum) for 1D tensor
- Add Cuda support
- Backward support
- Documentation for the op
- Perf optimizations and benchmark util
- Support for multi dimensional tensors (on data and lengths) (not high priority)
- Support for 'indices' (not high priority)

Test Plan: Added unit test

Differential Revision: D26952075

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -9289,3 +9289,8 @@
 - func: _test_ambiguous_defaults.b(Tensor dummy, int a=2, str b="2") -> Tensor
   cpp_no_default_args: ['a', 'b']
   python_module: nn
+
+- func: segment_reduce(Tensor data, str reduce, Tensor? lengths=None, Tensor? indices=None, int axis=0, bool unsafe=False) -> Tensor
```
Comment (ngimel) [inline aten/src/ATen/native/native_functions.yaml:9293]: you probably want all the args except `data` to be kwarg-only, so do it like
```
segment_reduce(Tensor data, *, str reduce ....)
```

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+Tensor _segment_reduce_cpu(
+    const Tensor& data,
+    std::string reduce,
+    const c10::optional<Tensor>& lengths,
+    const c10::optional<Tensor>& indices,
+    int64_t axis,
+    bool unsafe) {
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:24]: pytorch usually supports wrapping of dimensions, so e.g. 0 and -1 would be valid `axis` argument for a 1d tensor, so before checking `axis` do 
```
axis = maybe_wrap_dim(axis, data.ndim());
```

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  TORCH_CHECK(data.dim() == 1);
+  TORCH_CHECK(!data.is_cuda());
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:26]: this check is not needed, dispatcher will take care of this

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  TORCH_CHECK(data.numel() > 0);
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:27]: if possible, you should support empty tensors, by returning empty output tensor with the corresponding sizes.

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  const auto data_contig = data.contiguous();
+
+  int64_t batch_size = lengths_contig.numel();
+  auto output = at::empty({batch_size}, data.options());
+
+  const auto* lengths_data = lengths_contig.data_ptr<int64_t>();
+  if (!unsafe) {
+    int64_t sum = 0;
+    for (int64_t i = 0; i < batch_size; ++i) {
+      sum += lengths_data[i];
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:51]: you should also be checking that length_data[i] is not negative here.

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    TORCH_CHECK(sum == data.numel());
+  }
+
+  AT_DISPATCH_ALL_TYPES(
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:56]: ALL is not really ALL in pytorch, so if you want to support e.g. bfloat16 you'd need to add it separately with DISPATCH_ALL_TYPES_AND

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      data_contig.scalar_type(), "_segment_reduce_cpu", ([&]() {
+        auto* output_data = output.data_ptr<scalar_t>();
+        const auto* values_data = data_contig.data_ptr<scalar_t>();
+        int64_t k = 0;
+        for (int64_t i = 0; i < batch_size; ++i) {
+          scalar_t reduction = std::numeric_limits<scalar_t>::lowest();
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (values_data[k] > reduction) {
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:64]: this does not properly handle nan values, you need something like
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan

File: aten/src/ATen/native/SegmentReduce.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+              reduction = values_data[k];
+            }
+            k++;
+          }
+          if (lengths_data[i] > 0) {
+            output_data[i] = reduction;
+          } else {
+            output_data[i] = (scalar_t)0;
```
Comment (ngimel) [inline aten/src/ATen/native/SegmentReduce.cpp:72]: for max reduction, you can't reduce over 0-length segments, 0 is not a proper identity element for reduction. 0 length is an invalid input for a max reduction.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 2 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:25]: you meant "ndimension", right? SG! updated.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 4 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:27]: sg, will follow-up in future diff with test as well.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 5 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:52]: Also checking 0 as well now, to avoid issue below. So if unsafe is True, we will set to numerical limit based on reduction, otherwise check of valid lengths here. Let me know if this makes sense.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 7 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:72]: Thanks for the pointer. Updated. I think you meant "at::_isnan(data) ? reduction ..." instead, right?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (serhaty) [reply aten/src/ATen/native/native_functions.yaml:9300]: Is there an example I can follow for this? Also, where would the default be set in this approach?

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 7 above)
Comment (ngimel) [reply aten/src/ATen/native/SegmentReduce.cpp:72]: It should be `data`, not `reduction`, to properly propagate `nan` (also, will be good to add a test).

File: test/test_segment_reductions.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+
+class TestSegmentReductions(TestCase):
+    @onlyCPU
+    @dtypes(torch.half, torch.bfloat16, torch.float, torch.double)
+    def test_max_simple_1d(self, device, dtype):
+        lengths = torch.tensor([1, 2, 3], device=device)
+        data = torch.tensor([1, float("nan"), 3, 4, 5, 6], device=device, dtype=dtype)
+        expected_result = torch.tensor([1, 3, 6], device=device, dtype=dtype)
```
Comment (ngimel) [inline test/test_segment_reductions.py:19]: Typically, operations in pytorch propagate `nans` - do you need different behavior here?

**Review by ngimel — COMMENTED** [review]
Looks good, there's an issue of desired nan propagation behavior.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 7 above)
Comment (serhaty) [reply aten/src/ATen/native/SegmentReduce.cpp:72]: I actually updated the test in the PR with none value. I guess expected behaviour is not clear for me. If one of the values is None for a reduction, do we want to return None reduction or result based on remaining items within the segment?

I don't have preference one way or another but would like to keep similar convention as rest of PyTorch regarding none handling. If returning None is preferred, will update the test and PR.

File: test/test_segment_reductions.py (hunk 9 above)
Comment (serhaty) [reply test/test_segment_reductions.py:19]: replied in earlier comment. Will update the PR to return nan instead of reduction based on remaining elements.

File: aten/src/ATen/native/SegmentReduce.cpp (hunk 7 above)
Comment (ngimel) [reply aten/src/ATen/native/SegmentReduce.cpp:72]: Default is propagate nan values, that is, if nan is encountered in the segment, it is returned. But the expression should be as in my first comment to return nan if either of the args is nan
```
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan
```
If `data` is `nan` it is returned, otherwise if `reduction` is `nan` it will be returned as the first arg to `std::max`, otherwise max(reduction, data) is returned.

File: test/test_segment_reductions.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    TestCase,
+    run_tests,
+)
+
+
+class TestSegmentReductions(TestCase):
+    @onlyCPU
+    @dtypes(torch.half, torch.bfloat16, torch.float, torch.double)
+    def test_max_simple_1d(self, device, dtype):
+        lengths = torch.tensor([1, 2, 3], device=device)
```
Comment (xw285cornell) [inline test/test_segment_reductions.py:17]: usually it'll be a good idea to use hypothesis to test more cases w/ random number, and use numpy to write a ref implementation?

**Review by xw285cornell — COMMENTED** [review]
Nice :)

File: test/test_segment_reductions.py (hunk 10 above)
Comment (ngimel) [reply test/test_segment_reductions.py:17]: We typically don't use hypothesis in torch testsuite

Comment (xw285cornell) [reply test/test_segment_reductions.py:17]: I see, is it because of pulling extra dependencies? JUst try to think how to make the test stronger and cover corner cases; Maybe we can use torch.randint to randomize the data (btw in the past we do see some corner cases which fails the job, e.g. when input data is empty)

Comment (ngimel) [reply test/test_segment_reductions.py:17]: We usually specifically test for corner-case inputs instead of hoping that they'd be randomly generated

Comment (serhaty) [reply test/test_segment_reductions.py:17]: I still have todo to add more testing, documentation and benchmark in follow-up diffs. So this is still WIP.

Facebook:
Once we have high confidence on the op, only then will switch Dper3's LengthsMax and variations to segment_reduce.

Comment (serhaty) [reply test/test_segment_reductions.py:17]: I still have todo to add more testing, documentation and benchmark in follow-up diffs. So this is still WIP.

**Comment by ngimel** [comment]
This looks good, can you please rebase and reexport to get CI signal? There are 17 failing tests, and even though they are failing due to connection issues, we need to get signal from them.

**Review by ngimel — APPROVED** [review]
